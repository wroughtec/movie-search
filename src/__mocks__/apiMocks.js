export const searchMock = {
  page: 1,
  total_results: 140,
  total_pages: 7,
  results: [
    {
      vote_count: 9703,
      id: 11,
      video: false,
      vote_average: 8.2,
      title: 'Star Wars',
      popularity: 44.515,
      poster_path: '/btTdmkgIvOi0FFip1sPuZI2oQG6.jpg',
      original_language: 'en',
      original_title: 'Star Wars',
      genre_ids: [],
      backdrop_path: '/4iJfYYoQzZcONB9hNzg0J0wWyPH.jpg',
      adult: false,
      overview:
        'Princess Leia is captured and held hostage by the evil Imperial forces in their effort to take over the galactic Empire. Venturesome Luke Skywalker and dashing captain Han Solo team together with the loveable robot duo R2-D2 and C-3PO to rescue the beautiful princess and restore peace and justice in the Empire.',
      release_date: '1977-05-25'
    }
  ]
};

export const configMock = {
  images: {
    base_url: 'http://image.tmdb.org/t/p/',
    secure_base_url: 'https://image.tmdb.org/t/p/',
    backdrop_sizes: ['w300', 'w780', 'w1280', 'original'],
    logo_sizes: ['w45', 'w92', 'w154', 'w185', 'w300', 'w500', 'original'],
    poster_sizes: ['w92', 'w154', 'w185', 'w342', 'w500', 'w780', 'original'],
    profile_sizes: ['w45', 'w185', 'h632', 'original'],
    still_sizes: ['w92', 'w185', 'w300', 'original']
  },
  change_keys: []
};

export const genresMock = [{ id: 1, name: 'new' }];

export const movieDetails = {
  adult: false,
  backdrop_path: '/4iJfYYoQzZcONB9hNzg0J0wWyPH.jpg',
  belongs_to_collection: {
    id: 10,
    name: 'Star Wars Collection',
    poster_path: '/iTQHKziZy9pAAY4hHEDCGPaOvFC.jpg',
    backdrop_path: '/d8duYyyC9J5T825Hg7grmaabfxQ.jpg'
  },
  budget: 11000000,
  genres: [{ id: 12, name: 'Adventure' }, { id: 28, name: 'Action' }, { id: 878, name: 'Science Fiction' }],
  homepage: 'http://www.starwars.com/films/star-wars-episode-iv-a-new-hope',
  id: 11,
  imdb_id: 'tt0076759',
  original_language: 'en',
  original_title: 'Star Wars',
  overview:
    'Princess Leia is captured and held hostage by the evil Imperial forces in their effort to take over the galactic Empire. Venturesome Luke Skywalker and dashing captain Han Solo team together with the loveable robot duo R2-D2 and C-3PO to rescue the beautiful princess and restore peace and justice in the Empire.',
  popularity: 47.377,
  poster_path: '/btTdmkgIvOi0FFip1sPuZI2oQG6.jpg',
  production_companies: [
    { id: 1, logo_path: '/o86DbpburjxrqAzEDhXZcyE8pDb.png', name: 'Lucasfilm', origin_country: 'US' },
    { id: 25, logo_path: '/qZCc1lty5FzX30aOCVRBLzaVmcp.png', name: '20th Century Fox', origin_country: 'US' }
  ],
  production_countries: [{ iso_3166_1: 'US', name: 'United States of America' }],
  release_date: '1977-05-25',
  revenue: 775398007,
  runtime: 121,
  spoken_languages: [{ iso_639_1: 'en', name: 'English' }],
  status: 'Released',
  tagline: 'A long time ago in a galaxy far, far away...',
  title: 'Star Wars',
  video: false,
  vote_average: 8.2,
  vote_count: 9718
};
