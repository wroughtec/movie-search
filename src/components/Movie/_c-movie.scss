@import '../../scss/_mixins/m-container';

.c-movie {
  margin-top: -5vh;
}

.c-movie__container {
  margin-top: 5vh;
  margin-bottom: 5rem;

  @include container;
}

.c-movie__link {
  color: var(--btnBackground);
  text-decoration: none;

  &:hover {
    color: var(--btnHoverBackground);
  }
}

.c-movie__header {
  margin-bottom: 1rem;

  font-weight: 700;
  text-transform: uppercase;
}

.c-movie__title {
  margin-top: 0;
  margin-bottom: 0;

  font-size: 3rem;
}

.c-movie__tagline {
  margin-top: 0;
  margin-bottom: 0;

  font-size: 1.5rem;
}

.c-movie__content {
  display: grid;
  margin-top: 2rem;

  grid-gap: 2rem;
  grid-template-columns: repeat(auto-fit, minmax(30rem, 1fr));
}

.c-movie__meta {
  display: grid;
  padding: 2rem;
  margin-bottom: 2rem;

  grid-gap: 1.5rem;
  grid-template-columns: repeat(auto-fit, minmax(10rem, 1fr));

  background-color: var(--inputBackground);
  border-radius: 1rem;
}

.c-movie__heading {
  margin-top: 0;
  margin-bottom: 0;

  font-size: 2rem;
  font-weight: 400;
}
