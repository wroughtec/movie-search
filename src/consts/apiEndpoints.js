// @flow

export const searchEndpoint = 'search/movie';
export const configEndpoint = 'configuration';
export const genresEndpoint = 'genre/movie/list';
export const popularEndpoint = 'movie/popular';
export const movieDetailsEndpoint = (id: string) => `movie/${id}`;
