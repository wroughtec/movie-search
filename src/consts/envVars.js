// @flow

export const BASE_URL: ?string = process.env.API_URL;
export const KEY: ?string = process.env.API_KEY;
