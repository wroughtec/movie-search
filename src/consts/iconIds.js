// @flow

export const spinnerIcon = 'spinner';
