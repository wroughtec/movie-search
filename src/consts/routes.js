export const home = '/';
export const search = '/search/';
export const movie = '/movie/';
