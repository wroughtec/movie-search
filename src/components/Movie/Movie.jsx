// @flow

import React, { Component } from 'react';
import RequestMovies from 'utils/requestMovies';
import { produce } from 'immer';
import { Spinner } from 'components/Spinner/Spinner';
import { Poster } from 'components/Poster/Poster';
import { Link } from '@reach/router';
import Currency from 'react-currency-formatter';
import './_c-movie.scss';

type Props = {
  movieId: string,
  size: string,
  imageBaseUrl: string
};

type State = {
  loading: boolean,
  movieDetails: any
};

export class Movie extends Component<Props, State> {
  state = {
    loading: true,
    movieDetails: {}
  };

  componentDidMount() {
    const { movieId } = this.props;

    if (movieId) {
      this.getMovieDetails(movieId);
    }
  }

  getMovieDetails = async (id: string) => {
    const movie = await RequestMovies.movieDetails(id);

    if (movie && movie.id) {
      this.setState(
        produce(this.state, draft => {
          draft.movieDetails = movie;
          draft.loading = false;
        })
      );
    }
  };

  movieDetails = () => {
    const {
        movieDetails: {
          title,
          backdrop_path: path,
          overview,
          homepage,
          tagline,
          runtime,
          revenue,
          release_date: releaseDate,
          vote_average: rating,
          genres,
          production_companies: companies
        }
      } = this.state,
      { imageBaseUrl, size } = this.props;

    return (
      <article className="c-movie__details">
        <Poster imageBaseUrl={imageBaseUrl} size={size} path={path} title={title} />
        <div className="c-movie__container">
          <header className="c-movie__header">
            <h1 className="c-movie__title">
              {!homepage && title}
              {homepage && (
                <Link className="c-movie__link" to={homepage} target="_blank">
                  {title}
                </Link>
              )}
            </h1>
            {tagline && <h2 className="c-movie__tagline">{tagline}</h2>}
            {companies &&
              companies.length > 0 && (
                <div className="c-genre-tags">
                  {companies.map(company => {
                    const { id, name } = company;

                    return (
                      <span key={id} className={`c-genre-tags__tag c-genre-tags__tag--${id}`}>
                        {name}
                      </span>
                    );
                  })}
                </div>
              )}
          </header>
          <div className="c-movie__content">
            <div className="c-movie__description">
              <p className="c-movie__body">{overview}</p>
              {genres &&
                genres.length > 0 && (
                  <div className="c-genre-tags">
                    {genres.map(genre => {
                      // @TODO refactor to use the GenreTags componet
                      const { id, name } = genre;

                      return (
                        <span key={id} className={`c-genre-tags__tag c-genre-tags__tag--${id}`}>
                          {name}
                        </span>
                      );
                    })}
                  </div>
                )}
            </div>
            <div className="c-movie__meta">
              {releaseDate && (
                <div className="c-movie__group">
                  <h4 className="c-movie__heading">Release</h4>
                  <span className="c-movie__detail">{releaseDate}</span>
                </div>
              )}
              {runtime && (
                <div className="c-movie__group">
                  <h4 className="c-movie__heading">Length</h4>
                  <span className="c-movie__detail">{runtime} mins</span>
                </div>
              )}
              {revenue && (
                <div className="c-movie__group">
                  <h4 className="c-movie__heading">Box office</h4>
                  <span className="c-movie__detail">
                    <Currency quantity={revenue} currency="USD" />
                  </span>
                </div>
              )}
              {rating && (
                <div className="c-movie__group">
                  <h4 className="c-movie__heading">Rating</h4>
                  <span className="c-movie__detail">
                    {rating}
                    /10
                  </span>
                </div>
              )}
            </div>
          </div>
        </div>
      </article>
    );
  };

  render() {
    const { loading } = this.state;
    return (
      <section className="c-movie">
        {loading && <Spinner />}
        {!loading && this.movieDetails()}
      </section>
    );
  }
}
