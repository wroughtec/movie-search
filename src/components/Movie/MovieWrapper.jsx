// @flow

import React from 'react';
import { Consumer } from 'components/SearchContext/SearchContext';
import { Movie } from './Movie';

export const MovieWrapper = (props: any) => (
  <Consumer>
    {({ xl, imageBaseUrl }: SearchContextType) => <Movie {...props} size={xl} imageBaseUrl={imageBaseUrl} />}
  </Consumer>
);
