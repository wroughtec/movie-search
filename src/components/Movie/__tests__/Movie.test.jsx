// @flow

import React from 'react';
import { shallow, mount } from 'enzyme';
import fetchMock from 'fetch-mock';
import { Movie } from 'components/Movie/Movie';
import { BASE_URL, KEY } from 'consts/envVars';
import { movieDetails } from 'mocks/apiMocks';
import { MovieWrapper } from 'components/Movie/MovieWrapper';

fetchMock.get(`${BASE_URL}/movie/11?api_key=${KEY}`, JSON.stringify({ ...movieDetails }));

describe('Movie', () => {
  let component, props;

  beforeEach(() => {
    props = {
      movieId: '11',
      size: 'xl',
      imageBaseUrl: 'http://url.co/'
    };

    component = shallow(<Movie {...props} />);
  });

  test('Renders', () => {
    expect(component).toMatchSnapshot();
  });

  test('load data on componentDidMount', async () => {
    const getMovieDetails = jest.spyOn(component.instance(), 'getMovieDetails');
    await component.instance().componentDidMount();

    expect(getMovieDetails).toBeCalled();
  });

  test('no data componentDidMount', async () => {
    const newProps = { ...props, movieId: '' },
      newComponent = shallow(<Movie {...newProps} />);
    const getMovieDetails = jest.spyOn(newComponent.instance(), 'getMovieDetails');
    await component.instance().componentDidMount();

    expect(getMovieDetails).not.toBeCalled();
  });

  test('no data from api', async () => {
    fetchMock.config.overwriteRoutes = true;
    fetchMock.get(`${BASE_URL}/movie/11?api_key=${KEY}`, JSON.stringify({ error: 'error' }));
    const newComponent = shallow(<Movie {...props} />);

    expect(newComponent).toMatchSnapshot();
  });

  // @TODO look into better ways of testing the new Context API seems Enzymes support in not yet complete
  test('wrapper renders', () => {
    const wrapper = shallow(<MovieWrapper id="11" />, {
      context: {
        xl: 'xl',
        imageBaseUrl: 'http://url.co/'
      }
    });

    expect(wrapper).toMatchSnapshot();
  });
});
