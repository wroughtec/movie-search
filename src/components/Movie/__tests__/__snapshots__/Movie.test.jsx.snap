// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Movie
    imageBaseUrl="http://url.co/"
    movieId="11"
    size="xl"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Spinner />,
        false,
      ],
      "className": "c-movie",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Spinner />,
          false,
        ],
        "className": "c-movie",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Movie no data from api 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Movie
    imageBaseUrl="http://url.co/"
    movieId="11"
    size="xl"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Spinner />,
        false,
      ],
      "className": "c-movie",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Spinner />,
          false,
        ],
        "className": "c-movie",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Movie wrapper renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieWrapper
    id="11"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": Object {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_calculateChangedBits": null,
      "_currentRenderer": null,
      "_currentRenderer2": null,
      "_currentValue": Object {
        "genres": Array [],
        "handleMovieSubmit": [Function],
        "handleSearchChange": [Function],
        "imageBaseUrl": "",
        "lg": "",
        "loading": true,
        "md": "",
        "popularPageNo": 1,
        "popularResults": Object {},
        "searchPageNo": 1,
        "searchResults": Object {},
        "searchTerms": "",
        "sm": "",
        "updateSearchTerms": [Function],
        "xl": "",
      },
      "_currentValue2": Object {
        "genres": Array [],
        "handleMovieSubmit": [Function],
        "handleSearchChange": [Function],
        "imageBaseUrl": "",
        "lg": "",
        "loading": true,
        "md": "",
        "popularPageNo": 1,
        "popularResults": Object {},
        "searchPageNo": 1,
        "searchResults": Object {},
        "searchTerms": "",
        "sm": "",
        "updateSearchTerms": [Function],
        "xl": "",
      },
      "unstable_read": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "genres": Array [],
          "handleMovieSubmit": [Function],
          "handleSearchChange": [Function],
          "imageBaseUrl": "",
          "lg": "",
          "loading": true,
          "md": "",
          "popularPageNo": 1,
          "popularResults": Object {},
          "searchPageNo": 1,
          "searchResults": Object {},
          "searchTerms": "",
          "sm": "",
          "updateSearchTerms": [Function],
          "xl": "",
        },
        "_currentValue2": Object {
          "genres": Array [],
          "handleMovieSubmit": [Function],
          "handleSearchChange": [Function],
          "imageBaseUrl": "",
          "lg": "",
          "loading": true,
          "md": "",
          "popularPageNo": 1,
          "popularResults": Object {},
          "searchPageNo": 1,
          "searchResults": Object {},
          "searchTerms": "",
          "sm": "",
          "updateSearchTerms": [Function],
          "xl": "",
        },
        "unstable_read": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "context": Object {
      "imageBaseUrl": "http://url.co/",
      "xl": "xl",
    },
    "hydrateIn": undefined,
  },
}
`;
